File: internal/cmd/cards.go
package cmd

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/spf13/cobra"

	"mm-cli/internal/apps"
	"mm-cli/internal/card"
)

// NewCardsCmd builds `mm cards [<app>] [--refresh]`.
func NewCardsCmd() *cobra.Command {
	c := &cobra.Command{
		Use:     "cards [app]",
		Aliases: []string{"card"},
		Short:   "Capability matrix / per-app Agent Card",
		Args:    cobra.MaximumNArgs(1),
		RunE:    runCards,
	}
	c.Flags().Bool("refresh", false, "Bypass the 24h cache")
	return c
}

func NewManifestCmd() *cobra.Command {
	c := &cobra.Command{
		Use:   "manifest [app]",
		Short: "Wire-level manifest (deeper than the Card)",
		Args:  cobra.MaximumNArgs(1),
		RunE:  runManifest,
	}
	c.Flags().Bool("refresh", false, "Bypass the 24h cache")
	return c
}

func runCards(cmd *cobra.Command, args []string) error {
	if len(args) > 0 && args[0] == "help" {
		return cmd.Help()
	}
	refresh, _ := cmd.Flags().GetBool("refresh")
	wantJSON, _ := cmd.Root().PersistentFlags().GetBool("json")
	ctx := cmd.Context()

	if len(args) == 0 {
		// Capability matrix.
		type row struct {
			slug         string
			name         string
			capabilities []string
			tools        int
			err          error
		}
		rows := make([]row, 0, len(apps.Registry))
		for slug := range apps.Registry {
			c, err := card.Load(ctx, slug, refresh)
			r := row{slug: slug}
			if err != nil {
				r.err = err
			} else {
				r.name = c.Name
				r.capabilities = c.Capabilities
				r.tools = len(c.Tools)
			}
			rows = append(rows, r)
		}
		if wantJSON {
			out, _ := json.MarshalIndent(rows, "", "  ")
			fmt.Println(string(out))
			return nil
		}
		fmt.Println("# Apps")
		fmt.Println()
		for _, r := range rows {
			if r.err != nil {
				fmt.Printf("  %s  (card fetch failed: %s)\n", padRight(r.slug, 12), r.err)
				continue
			}
			caps := strings.Join(r.capabilities, ",")
			if caps == "" {
				caps = "—"
			}
			fmt.Printf("  %s  %s  caps=%-30s tools=%d\n",
				padRight(r.slug, 12), padRight(r.name, 22), caps, r.tools)
		}
		return nil
	}

	slug := args[0]
	c, err := card.Load(ctx, slug, refresh)
	if err != nil {
		return err
	}
	if wantJSON {
		out, _ := json.MarshalIndent(c, "", "  ")
		fmt.Println(string(out))
		return nil
	}
	fmt.Printf("# %s\n\n", c.Name)
	if c.Description != "" {
		fmt.Println(c.Description)
		fmt.Println()
	}
	if c.Version != "" {
		fmt.Printf("Version: %s\n", c.Version)
	}
	if len(c.Capabilities) > 0 {
		fmt.Printf("Capabilities: %s\n", strings.Join(c.Capabilities, ", "))
	}
	if c.MCPURL != nil && *c.MCPURL != "" {
		fmt.Printf("MCP: %s\n", *c.MCPURL)
	}
	if len(c.Tools) > 0 {
		fmt.Println()
		fmt.Println("## Tools")
		for _, t := range c.Tools {
			ann := []string{}
			if t.ReadOnlyHint != nil && *t.ReadOnlyHint {
				ann = append(ann, "read-only")
			}
			if t.DestructiveHint != nil && *t.DestructiveHint {
				ann = append(ann, "destructive")
			}
			tagPart := ""
			if len(ann) > 0 {
				tagPart = "  [" + strings.Join(ann, ", ") + "]"
			}
			fmt.Printf("  - %s%s\n", t.Name, tagPart)
			if t.Description != "" {
				fmt.Printf("    %s\n", t.Description)
			}
		}
	}
	if len(c.Aliases) > 0 {
		fmt.Println()
		fmt.Println("## Aliases")
		for verb, a := range c.Aliases {
			fmt.Printf("  %s → %s.%s\n", verb, a.Feature, a.Action)
		}
	}
	return nil
}

File: src/commands/cards.ts
// ⚠️ LEGACY TYPESCRIPT PORT — NOT the live `mm` binary. The live CLI is Go;
// this command lives in internal/cmd/. Editing this file changes nothing in
// `mm` (it only builds the separate, unused `mm-ts`). Fix the .go file instead.

/**
 * `mm cards` — discovery surface across all registered apps.
 *
 *   mm cards                    # capability matrix across all apps
 *   mm cards <slug>             # full Card for one app
 *   mm cards <slug> --json      # raw JSON
 *   mm cards [<slug>] --refresh # bypass 24h cache
 *
 * The Card is each app's `/.well-known/agent.json` — name, description,
 * capabilities, curated tools (with destructive/readonly hints), and
 * optional aliases. This is the agent-facing view; for the full wire
 * surface use `mm manifest <app>`.
 */

import { listApps, resolveApp } from '../apps';
import { loadAgentCard, type AgentCard } from '../agent-card';

const CAPS = ['ask', 'chat', 'search', 'writes'] as const;

function capRow(card: AgentCard | null, slug: string, url: string): string {
	const caps = card?.capabilities ?? [];
	const cells = CAPS.map((c) => (caps.includes(c) ? c : '·'.padEnd(c.length))).join('  ');
	const status = card ? cells : 'unreachable';
	const desc = card?.description ? ` — ${card.description.slice(0, 60)}` : '';
	return `  ${slug.padEnd(10)} ${status.padEnd(28)} ${url}${desc}`;
}

export async function cardsDispatch(
	args: string[],
	flags: { json?: boolean; refresh?: boolean }
) {
	const slug = args[0];

	if (!slug) {
		const apps = listApps();
		const results: { slug: string; url: string; card: AgentCard | null }[] = [];
		await Promise.all(
			apps.map(async (app) => {
				try {
					const card = await loadAgentCard(app.slug, { refresh: flags.refresh });
					results.push({ slug: app.slug, url: app.url, card });
				} catch {
					results.push({ slug: app.slug, url: app.url, card: null });
				}
			})
		);
		results.sort((a, b) => a.slug.localeCompare(b.slug));

		if (flags.json) {
			console.log(JSON.stringify(results, null, 2));
			return;
		}

		console.log(`Apps (${results.length}) — capability matrix\n`);
		console.log(`  ${'slug'.padEnd(10)} ${CAPS.join('  ').padEnd(28)} url`);
		for (const r of results) {
			console.log(capRow(r.card, r.slug, r.url));
		}
		console.log('');
		console.log('Use: mm cards <slug>          full Card for one app');
		console.log('     mm <slug> ask "..."      ask the app a question (agent.chat)');
		console.log('     mm <slug> find "..."     search the app (agent.search)');
		return;
	}

	resolveApp(slug); // throws on unknown
	const card = await loadAgentCard(slug, { refresh: flags.refresh });

	if (flags.json) {
		console.log(JSON.stringify(card, null, 2));
		return;
	}

	console.log(`${card.name}${card.version ? ` (v${card.version})` : ''}`);
	if (card.description) console.log(`  ${card.description}`);
	console.log('');
	console.log(`  capabilities  ${(card.capabilities ?? []).join(', ') || '(none)'}`);
	if (card.chatUrl) console.log(`  chat          ${card.chatUrl}`);
	if (card.mcpUrl) console.log(`  mcp           ${card.mcpUrl}`);
	if (card.auth?.length) console.log(`  auth          ${card.auth.join(', ')}`);

	if (card.tools?.length) {
		console.log(`\nTools (${card.tools.length}):`);
		for (const t of card.tools) {
			const hints: string[] = [];
			if (t.readOnlyHint) hints.push('read-only');
			if (t.destructiveHint) hints.push('destructive');
			if (t.idempotentHint) hints.push('idempotent');
			const tag = hints.length ? `  [${hints.join(', ')}]` : '';
			console.log(`  ${t.name}${tag}`);
			if (t.description) console.log(`    ${t.description}`);
		}
	}

	if (card.aliases && Object.keys(card.aliases).length) {
		console.log(`\nAliases:`);
		for (const [verb, target] of Object.entries(card.aliases)) {
			const desc = target.description ? ` — ${target.description}` : '';
			console.log(`  mm ${slug} ${verb.padEnd(12)}  → ${target.feature}.${target.action}${desc}`);
		}
	}

	console.log('');
	console.log(`Wire surface: mm manifest ${slug}`);
}